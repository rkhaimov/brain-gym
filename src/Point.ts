import { assertNotUndefined } from './utils';

export class Point {
  constructor(public position: Position) {}

  getAngleBetween(point: Point) {
    return this.atan2(
      point.position.y - this.position.y,
      point.position.x - this.position.x
    );
  }

  move(position: Position) {
    this.position = position;
  }

  getDistanceFrom(point: Point) {
    return Math.sqrt(
      Math.pow(this.position.x - point.position.x, 2) +
        Math.pow(this.position.y - point.position.y, 2)
    );
  }

  moveAndRender(node: HTMLDivElement, position: Position) {
    this.move(position);

    node.style.left = `${this.position.x}px`;
    node.style.top = `${this.position.y}px`;
  }

  render() {
    const node = document.createElement('div');

    node.classList.add('point');

    const root = document.querySelector('#root');

    assertNotUndefined(root);

    root.appendChild(node);

    node.style.left = `${this.position.x}px`;
    node.style.top = `${this.position.y}px`;

    return node;
  }

  private atan2(y: number, x: number) {
    if (x < 0) {
      return Math.atan(y / x) + Math.PI;
    }

    if (y < 0) {
      return Math.atan(y / x) + 2 * Math.PI;
    }

    return Math.atan(y / x);
  }
}

type Position = { x: number; y: number };
