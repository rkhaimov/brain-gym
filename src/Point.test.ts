import { Point } from './Point';

describe('Point', () => {
  it('should return angle', () => {
    const p1 = new Point({ x: 2, y: 2 }, noop);
    const p2 = new Point({ x: 4, y: 2 }, noop);

    expect(p1.getAngleBetween(p2)).toBe(0);

    p2.move({ x: 2, y: 4 });

    expect(p1.getAngleBetween(p2)).toBe(Math.PI / 2);

    p2.move({ x: 0, y: 2 });

    expect(p1.getAngleBetween(p2)).toBe(Math.PI);

    p2.move({ x: 2, y: 0 });

    expect(p1.getAngleBetween(p2)).toBe((3 * Math.PI) / 2);

    p2.move({ x: 3, y: 1 });

    expect(p1.getAngleBetween(p2)).toMatchSnapshot();
  });

  it('should return distance', () => {
    const p1 = new Point({ x: 0, y: 0 }, noop);
    const p2 = new Point({ x: 3, y: 4 }, noop);

    expect(p1.getDistanceFrom(p2)).toBe(5);

    p1.move({ x: 2, y: 2 });
    p2.move({ x: 2, y: 2 });

    expect(p1.getDistanceFrom(p2)).toBe(0);

    p2.move({ x: 3, y: 1 });

    expect(p1.getDistanceFrom(p2)).toBe(Math.sqrt(2));
  });
});

function noop() {}
