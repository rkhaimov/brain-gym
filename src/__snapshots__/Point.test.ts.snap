// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Point should return angle 1`] = `5.497787143782138`;
